### ISSUE TITLE
Aborter throws timeout error even though it succeeds on download of BlockBlobUrl

### ISSUE THREAD

**Issue by Ventis** [Azure/azure-storage-js#67]
### Which service(blob, file, queue, table) does this issue concern?
Blob Storage

### Which version of the SDK was used?
10.3.0

### What's the Node.js/Browser version?
NodeJS 11.8

### What problem was encountered?
Aborter times out after BlockBlobURL download method has already completed:

```
 Error: The request was aborted
     at new RestError (/shared/node_modules/@azure/ms-rest-js/dist/msRest.node.js:1397:28)
     at a.<anonymous> (/shared/node_modules/@azure/storage-blob/dist/index.js:1:11269)
     at /shared/node_modules/@azure/storage-blob/dist/index.js:1:1277
    at Array.forEach (<anonymous>)
    at a.abort (/shared/node_modules/@azure/storage-blob/dist/index.js:1:1255)
     at Timeout.<anonymous> (/shared/node_modules/@azure/storage-blob/dist/index.js:1:519)
     at listOnTimeout (timers.js:324:15)
    at processTimers (timers.js:268:5)
```

### Steps to reproduce the issue?

My Azure storage utils class AzureStorage.ts:
```
import {
  BlockBlobURL,
  uploadFileToBlockBlob,
  Aborter,
  SharedKeyCredential,
  ContainerURL,
  ServiceURL,
  StorageURL
} from '@azure/storage-blob';
const path = require('path');

const ONE_MINUTE = 60 * 1000;

class AzureStorage {
  credentials: SharedKeyCredential;
  accountName: string;
  timeout: number;
  serviceURL: ServiceURL;
  pipeline: any;

  constructor(accountName: string, accessKey: string, timeout: number = 2) {
    this.credentials = new SharedKeyCredential(accountName, accessKey);
    this.accountName = accountName;
    this.timeout = timeout;
    this.createPipeline = this.createPipeline.bind(this);
    this.createServiceURL = this.createServiceURL.bind(this);
    this.createContainerURL = this.createContainerURL.bind(this);
    this.createContainer = this.createContainer.bind(this);
    this.getContainerNames = this.getContainerNames.bind(this);
    this.uploadLocalFile = this.uploadLocalFile.bind(this);
    this.downloadFile = this.downloadFile.bind(this);
  }

  createPipeline() {
    this.pipeline = StorageURL.newPipeline(this.credentials);
    return this.pipeline;
  }

  createServiceURL() {
    this.serviceURL = new ServiceURL(
      `https://${this.accountName}.blob.core.windows.net`,
      this.createPipeline()
    );
    return this.serviceURL;
  }

  createContainerURL(containerName: string) {
    return ContainerURL.fromServiceURL(this.createServiceURL(), containerName);
  }

  async createContainer(containerName: string) {
    await this.createContainerURL(containerName).create(this.createAborter());
  }

  async getContainerNames() {
    let marker;
    return await this.createServiceURL().listContainersSegment(
      this.createAborter(),
      marker
    );
  }

  checkIfContainerExists(containerName: string, containers: any) {
    if (!containers) return false;
    const found = containers.find(
      container => containerName === container.name
    );
    return found !== undefined;
  }

  async uploadLocalFile(containerName: string, filePath: string) {
    filePath = path.resolve(filePath);
    const fileName = path.basename(filePath);
    const blockBlobURL = BlockBlobURL.fromContainerURL(
      this.createContainerURL(containerName),
      fileName
    );
    return await uploadFileToBlockBlob(this.createAborter(), filePath, blockBlobURL);
  }

  async downloadFile(containerName: string, fileName: string) {
    const blockBlobURL = BlockBlobURL.fromContainerURL(
      this.createContainerURL(containerName),
      fileName
    );
    const downloadResponse = await blockBlobURL.download(this.createAborter(), 0);
    return downloadResponse.readableStreamBody;
  }

  createAborter() {
    return Aborter.timeout(this.timeout * ONE_MINUTE);
  }
}

export default AzureStorage;

```

When I use above class' downloadFile method to download a blob from BlobStorage, it succeeds and returns the readableStreamBody but the Aborter still throws the timeout error after 1 minute.

### Have you found a mitigation/solution?
Using Aborter.none. But then you don't have a working timeout when it's needed ;)

**Comment by Ventis** [comment]
This doesn't happen with uploadFileToBlockBlob btw.

**Comment by XiaoningLiu** [comment]
@Ventis This is a bug in `RetriableReadableStream` which used by `blockBlobURL.download`. `RetriableReadableStream` is referenced by the aborter timer.

Before we have a fix, you can manually cancel the Aborter timer when you finish the download by:
`aborter.cancelTimer()`.

**Comment by ricklancee** [comment]
We had this issue on reading that the Aborter timer would not be cleared; and since the try catch around our filesystem methods would not catch the internal errors throw inside aborter's timeoutwe handled the timeout ourself, like so: 

```
// example filesystem
class FileSystem {
    private aborterTimeoutSeconds: number = 60

    public async read(filePath: string): Promise<Readable> {
        const blobUrl = this.getBlobUrl(filePath)

        return new Promise<Readable>(async (resolve, reject) => {
            try {
                const { aborter, clearAborterTimer } = this.getNewAborter(reject)
                
                const downloadBlockBlobResponse = await blobUrl.download(aborter, 0)

                downloadBlockBlobResponse.readableStreamBody.on('end', () => {
                    clearAborterTimer()
                })

                // Use PassThrough to turn the original stream into Readable
                const passthrough = new PassThrough()
                downloadBlockBlobResponse.readableStreamBody.pipe(passthrough)

                resolve(passthrough)
            } catch (e) {
                reject(e)
            }
        })
    }

    private getNewAborter(reject: Function) {
        const aborter = Aborter.none

        const timer = setTimeout(() => {
            try {
                aborter.abort()
            } catch (err) {
                reject(err)
            }
        }, this.aborterTimeoutSeconds * 1000)

        const clearAborterTimer = () => {
            clearTimeout(timer)
        }

        return {
            aborter,
            clearAborterTimer,
        }
    }
}
```

**Comment by NicholasJarr** [comment]
It is happenning in uploadStreamToBlockBlob too. I'm using version 10.3.0

**Comment by NicholasJarr** [comment]
```aborter.cancelTimer()``` worked for me

**Comment by olad32** [comment]
Unfortunately for Typescript users cancelTimer() is a private method of Aborter class, so this "hack" is needed : 
`Aborter.prototype['cancelTimer'].apply(aborter);`

**Comment by XiaoningLiu** [comment]
Then we need to make `cancelTimer` to public

**Comment by XiaoningLiu** [comment]
This is fixed in 10.4.0, see https://www.npmjs.com/package/@azure/storage-blob

**Comment by highfield** [comment]
@XiaoningLiu did you patch the Typescript for bringing public `cancelTimer`? I'm using the release 10.5, but it seems still marked as private.

**Comment by XiaoningLiu** [comment]
@highfield We do not need to call `cancelTimer` as SDK fixed the issue and will not throw timeout error after download success.

**Comment by highfield** [comment]
@XiaoningLiu I wish to believe you, but I had the same problem as described above. Just after a short time the error crashed my app. That was yesterday afternoon. 

Once I added the below workaround, the things seem going fine (no error after 12 hours):

```
        const aborter = Aborter.timeout(30 * ONE_MINUTE);
        const blockBlobURL = BlockBlobURL.fromContainerURL(this.containerURL, actualPath);
        const response = await blockBlobURL.download(aborter, 0);
        (aborter as any).cancelTimer();   // this makes the app working fine
        return response.readableStreamBody as stream.Readable;

```

**Comment by XiaoningLiu** [comment]
@highfield Sorry for that. Please leverage the workaround temporiaily.
@ljian3377 Can you help quick check this issue still reproduable in latest v10 & v12 blob package versions?

**Comment by ljian3377** [comment]
> @highfield Sorry for that. Please leverage the workaround temporiaily.
> @ljian3377 Can you help quick check this issue still reproduable in latest v10 & v12 blob package versions?

Reproduced in 10.3 and 10.4, but not in v12.
It seems to be fixed in @azure/storage-blob version: 12.0.0. And the API has some changes. `Aborter` now goes in `AbortController`. And `blockBlobURL` basically is replaced by `BlockBlobClient`.

Code snippet used to reproduce:
```
const downloadBlockBlobResponse: BlobDownloadResponseModel = await blockBlobClient.download(0, undefined, { abortSignal: AbortController.timeout(60000) });

setTimeout(()=>console.log("wait done"), 100000);
```

**Comment by Mitsuo-Yoshida** [comment]
I know this repo is no longer maintained 
But mentioning it in this issue because I found how to fix this bug here
> Reproduced in 10.3 and 10.4, but not in v12.
> It seems to be fixed in @azure/storage-blob version: 12.0.0. And the API has some changes. `Aborter` now goes in `AbortController`. And `blockBlobURL` basically is replaced by `BlockBlobClient`.
> 
> Code snippet used to reproduce in v12:
> 
> ```
> const downloadBlockBlobResponse: BlobDownloadResponseModel = await blockBlobClient.download(0, undefined, { abortSignal: AbortController.timeout(60000) });
> 
> setTimeout(()=>console.log("wait done"), 100000);
> ```
I reproduced this bug in @azure/storage-blob version: **12.0.0** and **12.1.2** using the code snippet provided by @ljian3377

So I used the workaround provided to fix the issue
> `Aborter.prototype['cancelTimer'].apply(aborter);`

**Comment by ljian3377** [comment]
@Mitsuo-Yoshida 
That's wierd. I still can't reproduce it. Could you please share more information of your reproduction? What envirornment?
Has the download completed when the timeout aborter aborts?
